### ISSUE TITLE
Environment_Enginer: Get Environment Panel Geometry

### ISSUE THREAD

**Issue by michaldengusiak** [BHoM/BHoM_Engine#965]
**GH workflow** 

Currently to get geometry - Polyline  from Panel we have three options: Method 1,2,3
See image below:
Method 1 - works fine but we need to use lots of steps, explode geometry get edges, get BHoM curves then transform to GH Curves and then GH Lines and then generate BHoM Polyline. 
Method 2 - we query EnvironmentObject  to get BHoM polyline bit this one is invalid, it is closed but due to many duplicate points does not return clean polyline
Method 3 - we can get Invalid BHoM polyline, query control points, delete duplicate points, create BHoM polyline, use GH Polyline and set to true to close it, 

All of steps are quite a long work around...i
Q1: Is there better way to do it I am now aware of?
Q2. Can we clean `Convert to Polyline` for  EnvironmentObject, to give us clean, closed polyline
Q3. For Dynamo users can we have `Convert to Line` for  EnvironmentObject
Q4. Can we modify `Create Polyline` to accept list of `Line-Like Curves` 
Q5. Can we modify `Create Polyline` to accept list of `Environment Edges`
![image](https://user-images.githubusercontent.com/26113670/58075572-f0b63400-7b9f-11e9-8876-37e2805be583.png)

**Comment by IsakNaslundBh** [comment]
Can give my point of view as tagged in the question:

> Q1: Is there better way to do it I am now aware of?
Not really right now though to bugs/problems in some methods. See next answers

> Q2. Can we clean Convert to Polyline for EnvironmentObject, to give us clean, closed polyline
As discussed this is a bug that needs to be fixed. this should be the goto solution

> Q3. For Dynamo users can we have Convert to Line for EnvironmentObject
I assume that the polyline method for a panel, once fixed, would be turned into a polycurve in dynamo? Would this be enough?

> Q4. Can we modify Create Polyline to accept list of Line-Like Curves
This is a rhino convert-issue I have experienced before.. @epignatelli did we get anywhere with methods wanting lines to accept the line-like curve? Think we had a look in the nurbs method?

> Q5. Can we modify Create Polyline to accept list of Environment Edges
Fundamentally, no. We should not have Create method for geometry from any of the object models. This whole issue should be solved by fixing Q2, which should be turned into a Query method for the panels/openings.

**Comment by FraserGreenroyd** [comment]
My answers below.

> Q1: Is there better way to do it I am now aware of?

The best way of doing this is querying the `EnvironmentObject` to get its `Polyline` representation. As @IsakNaslundBh noted with you yesterday, there is a bug with that component in duplicating points producing invalid `Polyline` representations.

> Q2. Can we clean Convert to Polyline for EnvironmentObject, to give us clean, closed polyline

This is indeed something which needs to be done and to me is the bug to be resolved. I will get a fix pushed up within the next half hour for you to test @michaldengusiak 

> Q3. For Dynamo users can we have Convert to Line for EnvironmentObject

I concur with @IsakNaslundBh - when Q2 is fixed this should be resolved for Dynamo.

> Q4. Can we modify Create Polyline to accept list of Line-Like Curves

No answer from me, that is for @epignatelli to comment on.

> Q5. Can we modify Create Polyline to accept list of Environment Edges

I agree with @IsakNaslundBh - the hierarchy is object models know about geometry, but geometry does not know about object models, and this should be the way forward. It is down to individual discipline engines to create/convert/query geometry from their object models as necessary.


To summarise then, Q2 is a bug which needs resolving which I will push a fix for shortly :smile:

**Comment by michaldengusiak** [comment]
ok this is now solved and script looks much neater
![image](https://user-images.githubusercontent.com/26113670/58093891-6386d580-7bc7-11e9-923e-7d19ec11ac7f.png)
